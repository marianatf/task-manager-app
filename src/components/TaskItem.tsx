import React, { useState } from "react";
import type { TaskItem as TaskItemType } from "../types";
import { priorityColor } from "../utils/priorityColor";

interface Props extends TaskItemType {
  onDelete: (id: string) => void;
  onToggleComplete: (id: string, completed: boolean) => void;
}

export const TaskItem: React.FC<Props> = ({
  id,
  completed,
  title,
  description,
  priority = "low",
  onDelete,
  onToggleComplete,
}) => {
  const [showDescription, setShowDescription] = useState(false);

  return (
    <div className="py-2 group">
      <div className="flex items-center justify-between">
        <div
          className="flex items-center gap-3 flex-grow cursor-pointer"
          onClick={() => setShowDescription((prev) => !prev)}
        >
          <input
            id={id}
            checked={completed}
            onChange={() => onToggleComplete(id, completed)}
            type="checkbox"
            className="w-5 h-5 accent-black"
            onClick={(e) => e.stopPropagation()}
          />
          <div className="flex flex-col">
            <div
              className={`font-semibold text-gray-900${
                completed ? " line-through" : ""
              }`}
            >
              {title}
            </div>
            {showDescription && (
              <div
                className={`text-gray-600 text-sm mt-1${
                  completed ? " line-through" : ""
                }`}
              >
                {description}
              </div>
            )}
          </div>
        </div>
        <div className="flex items-center gap-2 ml-4">
          <span
            className={`w-4 h-4 rounded-full ${priorityColor(priority)}`}
          ></span>
          <button
            className="flex items-center justify-center text-gray-400 hover:text-red-500 text-xl opacity-0 group-hover:opacity-100 transition-opacity"
            onClick={() => onDelete(id)}
          >
            &times;
          </button>
        </div>
      </div>
    </div>
  );
};
