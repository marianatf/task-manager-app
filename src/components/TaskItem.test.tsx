import { TaskItem } from "./TaskItem";
import { render, screen, fireEvent } from "@testing-library/react";
import { describe, expect, it, vi } from "vitest";

describe("TaskItem", () => {
  const mockOnDelete = vi.fn();
  const mockOnToggleComplete = vi.fn();

  it("should match snapshot", () => {
    const { asFragment } = render(
      <TaskItem
        onDelete={mockOnDelete}
        onToggleComplete={mockOnToggleComplete}
        id={""}
        title={""}
        completed={false}
        description={""}
        priority={"high"}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should call onDelete when delete button is clicked", () => {
    render(
      <TaskItem
        onDelete={mockOnDelete}
        onToggleComplete={mockOnToggleComplete}
        id={"1"}
        title={"test"}
        completed={false}
        description={""}
        priority={"medium"}
      />
    );

    const deleteBtn = screen.getByRole("button", { name: /Ã—/i });
    fireEvent.click(deleteBtn);
    expect(mockOnDelete).toHaveBeenCalledWith("1");
  });

  it("should call onToggleComplete when checkbox is clicked", () => {
    render(
      <TaskItem
        onDelete={mockOnDelete}
        onToggleComplete={mockOnToggleComplete}
        id={"2"}
        title={"test"}
        completed={false}
        description={""}
        priority={"low"}
      />
    );

    const checkbox = screen.getByRole("checkbox");
    fireEvent.click(checkbox);
    expect(mockOnToggleComplete).toHaveBeenCalledWith("2", false);
  });
});
