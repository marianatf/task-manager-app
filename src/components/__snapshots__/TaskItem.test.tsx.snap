// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TaskItem > should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="py-2 group"
  >
    <div
      class="flex items-center justify-between"
    >
      <div
        class="flex items-center gap-3 flex-grow cursor-pointer"
      >
        <input
          class="w-5 h-5 accent-black"
          id=""
          type="checkbox"
        />
        <div
          class="flex flex-col"
        >
          <div
            class="font-semibold text-gray-900"
          />
        </div>
      </div>
      <div
        class="flex items-center gap-2 ml-4"
      >
        <span
          class="w-4 h-4 rounded-full bg-red-200"
        />
        <button
          class="flex items-center justify-center text-gray-400 hover:text-red-500 text-xl opacity-0 group-hover:opacity-100 transition-opacity"
        >
          Ã—
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
